pkg/policyfilter/namespace.go

```diff
@@ -0,0 +1,137 @@
+type NamespaceMap struct {
+	cgroupIdMap *ebpf.Map
+	nsIdMap     *lru.Cache[uint64, nsId]
+	nsNameMap   *lru.Cache[nsId, uint64]
+	id          uint64
+}
+
+var global *NamespaceMap
+
+// newNamespaceMap returns a new namespace mapping. The namespace map consists of
+// two pieces. First a cgroup to ID map. The ID is useful for BPF so we can avoid
+// strings in BPF side. Then a stable ID to namespace mapping.
+func newNamespaceMap() (*NamespaceMap, error) {
+	if global != nil {
+		return global, nil
+	}
+
+	idCache, err := lru.New[uint64, nsId](namespaceCacheSize)
+	if err != nil {
+		return nil, fmt.Errorf("create namespace ID cache failed")
+	}
+	nameCache, err := lru.New[nsId, uint64](namespaceCacheSize)
+	if err != nil {
+		return nil, fmt.Errorf("create namespace name cache failed")
+	}
+
+	objName := config.ExecObj()
+	objPath := path.Join(option.Config.HubbleLib, objName)
+	spec, err := ebpf.LoadCollectionSpec(objPath)
+	if err != nil {
+		return nil, fmt.Errorf("loading spec for %s failed: %w", objPath, err)
+	}
+	nsMapSpec, ok := spec.Maps[CgrpNsMapName]
+	if !ok {
+		return nil, fmt.Errorf("%s not found in %s", CgrpNsMapName, objPath)
+	}
+
+	ret, err := ebpf.NewMap(nsMapSpec)
+	if err != nil {
+		return nil, err
+	}
+
+	mapDir := bpf.MapPrefixPath()
+	pinPath := filepath.Join(mapDir, CgrpNsMapName)
+	os.Remove(pinPath)
+	os.Mkdir(mapDir, os.ModeDir)
+	err = ret.Pin(pinPath)
+	if err != nil {
+		ret.Close()
+		return nil, fmt.Errorf("failed to pin Namespace map in %s: %w", pinPath, err)
+	}
+
+	global = &NamespaceMap{
+		cgroupIdMap: ret,
+		nsIdMap:     idCache,
+		nsNameMap:   nameCache,
+		id:          1,
+	}
+	return global, err
+}
+
+func GetNamespaceWorkloadFromId(id uint64) (string, string, bool) {
+	if global == nil {
+		return "", "", false
+	}
+
+	if ns, ok := global.nsIdMap.Get(id); ok {
+		return ns.namespace, ns.workload, true
+	}
+	return "", "", false
+}
+
+// addCgroupIDs add cgroups ids to the policy map
+// todo: use batch operations when supported
+func addCgroupIDs(cinfo []containerInfo, pod *podInfo) error {
+	m := global
+	for _, c := range cinfo {
+		key := nsId{
+			namespace: pod.namespace,
+			workload: pod.workload,
+		}
+		id, ok := m.nsNameMap.Get(key)
+		if ok == true {
+			if err := m.cgroupIdMap.Update(&c.cgID, id, ebpf.UpdateAny); err != nil {
+				logger.GetLogger().WithError(err).Warn("Unable to assign cgroup to existing namespace")
+			}
+			continue
+		}
+		logger.GetLogger().WithField("cgrp", c).WithField("pod", pod).WithField("id", m.id).Warn("update cgroupid map")
+
+		// If this is a new namespace we create a new map entry and bind it to a stable id.
+		if err := m.cgroupIdMap.Update(&c.cgID, m.id, ebpf.UpdateAny); err != nil {
+			logger.GetLogger().WithError(err).WithField("cgid", c.cgID).WithField("id", m.id).WithField("ns", c.name).Warn("Unable to insert cgroup id map")
+			continue
+		}
+		logger.GetLogger().Warn("nsIdMap.Add(%d %s\n", m.id, pod.namespace)
```

Comment:
Should this maybe be an `Info` or `Debug` instead of `Warn`?

Reply:
I'm going to remove it. I'll add a follow up tetra command to produce some debug tables that are more useful.